server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2229,64 +2267,60 @@ public void acquireReplicaOperationPermit(final long operationPrimaryTerm, final
                         shardState != IndexShardState.STARTED) {
                         throw new IndexShardNotStartedException(shardId, shardState);
                     }
-                    try {
-                        indexShardOperationPermits.blockOperations(30, TimeUnit.MINUTES, () -> {
-                            assert operationPrimaryTerm > primaryTerm :
-                                "shard term already update.  op term [" + operationPrimaryTerm + "], shardTerm [" + primaryTerm + "]";
-                            primaryTerm = operationPrimaryTerm;
-                            updateGlobalCheckpointOnReplica(globalCheckpoint, "primary term transition");
-                            final long currentGlobalCheckpoint = getGlobalCheckpoint();
-                            final long localCheckpoint;
-                            if (currentGlobalCheckpoint == SequenceNumbers.UNASSIGNED_SEQ_NO) {
-                                localCheckpoint = SequenceNumbers.NO_OPS_PERFORMED;
-                            } else {
-                                localCheckpoint = currentGlobalCheckpoint;
-                            }
-                            logger.trace(
+
+                    if (opPrimaryTerm > pendingPrimaryTerm) {
+                        bumpPrimaryTerm(opPrimaryTerm, () -> {
+                            // a primary promotion, or another primary term transition, might have been triggered concurrently to this
+                            // recheck under the operation permit if we can skip doing this work
+                            if (opPrimaryTerm == pendingPrimaryTerm) {
```

Comment:
how is that possible? shouldn't the pending primary term update under mutex prevent this?

Reply:
I'm not sure how it can "prevent" this.

Assume you have a replica term bump followed by a promotion to primary. The replica term bump will call asyncBlockOperations to run the above code. Assume that acquireReplicaOperationPermit leaves the mutex before the code in the operation block gets to execute. Then a primary promotion comes in from the cluster state, updating the pendingPrimaryTerm again. Note that there's a test for this (`testReplicaTermIncrementWithConcurrentPrimaryPromotion`).

Reply:
+1 to the test.

So say the increase of the term comes in first. That acquires the mutex, bumps the `pendingPrimaryTerm` and submits an async block under mutex.

Then the replica gets promoted. That means the term is higher. When the cluster state comes in, it bumps the `pendingPrimaryTerm` again and submit an async block. We know that they first async block code will run first, followed by the second, and I think that's OK.

If the reverse happens - i.e., the updateShardState comes in first, it will bump the pendingPrimaryTerm to a higher number than the replication operation, which will prevent the replication operation from submitting it's async block, so we're good here too. 

What am I missing?

Reply:
> We know that they first async block code will run first?

No we don't know that. See implementation of asyncBlockOperations. Both just submit a task to the generic threadpool. Both potentially race to the call of `doBlockOperations`